docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,1063 @@
+
+   // MaxConcurrentReconciles is the maximum number of concurrent Reconciles
+   // which can be run. Defaults to 10.
+   // +optional
+   MaxConcurrentReconciles *int `json:"maxConcurrentReconciles,omitempty"`
+
+   // Verbosity set the logs verbosity. Defaults to 1.
+   // Controller Manager flag is --verbosity.
+   // +optional
+   Verbosity int `json:"verbosity,omitempty"`
+
+   // Debug, if set, will override a set of fields with opinionated values for
+   // a debugging session. (Verbosity=5, ProfilerAddress=localhost:6060)
+   // +optional
+   Debug bool `json:"debug,omitempty"`
+
+   // FeatureGates define provider specific feature flags that will be passed
+   // in as container args to the provider's controller manager.
+   // Controller Manager flag is --feature-gates.
+   FeatureGates map[string]bool `json:"featureGates,omitempty"`
+}
+
+// DeploymentSpec defines the properties that can be enabled on the Deployment for the provider.
+type DeploymentSpec struct {
+   // Number of desired pods. This is a pointer to distinguish between explicit zero and not specified. Defaults to 1.
+   // +optional
+   Replicas *int `json:"replicas,omitempty"`
+
+   // NodeSelector is a selector which must be true for the pod to fit on a node.
+   // Selector which must match a node's labels for the pod to be scheduled on that node.
+   // More info: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
+   // +optional
+   NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+   // If specified, the pod's tolerations.
+   // +optional
+   Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+   // If specified, the pod's scheduling constraints
+   // +optional
+   Affinity *corev1.Affinity `json:"affinity,omitempty"`
+
+   // List of containers specified in the Deployment
+   // +optional
+   Containers []ContainerSpec `json:"containers"`
+}
+
+// ContainerSpec defines the properties available to override for each
+// container in a provider deployment such as Image and Args to the container’s
+// entrypoint.
+type ContainerSpec struct {
+   // Name of the container. Cannot be updated.
+   Name string `json:"name"`
+
+   // Container Image Name
+   // +optional
+   Image *ImageMeta `json:"image,omitempty"`
+
+   // Args represents extra provider specific flags that are not encoded as fields in this API.
+   // Explicit controller manager properties defined in the `Provider.ManagerSpec`
+   // will have higher precedence than those defined in `ContainerSpec.Args`.
+   // For example, `ManagerSpec.SyncPeriod` will be used instead of the
+   // container arg `--sync-period` if both are defined.
+   // The same holds for `ManagerSpec.FeatureGates` and `--feature-gates`.
+   // +optional
+   Args map[string]string `json:"args,omitempty"`
+
+   // List of environment variables to set in the container.
+   // +optional
+   Env []corev1.EnvVar `json:"env,omitempty"
```

Comment:
I added `Env` to explicitly support environment variables to the container. However as I was looking at the corev1 api I noticed the following comments and tags and I was curious if we need to specify these tags as well.

```
// +patchMergeKey=name
// +patchStrategy=merge
Env []EnvVar `json:"env,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,7,rep,name=env"`
```

Reply:
We should try them out, IIRC latest controller runtime/tools versions should support these, although it needs to be tested

Reply:
I'll leave out the annotations from the proposal for now and we can try adding them in when we start implementing.